File: app/src/main/java/com/example/pico_botella/viewmodel/GameViewModel.java
package com.example.pico_botella.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.pico_botella.model.GameChallengeUiState;
import com.example.pico_botella.model.Reto;
import com.example.pico_botella.repository.PokemonRepository;
import com.example.pico_botella.repository.RetoRepository;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GameViewModel extends AndroidViewModel {

    private final RetoRepository retoRepository;
    private final PokemonRepository pokemonRepository;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final MutableLiveData<GameChallengeUiState> challengeState = new MutableLiveData<>();

    public GameViewModel(@NonNull Application application) {
        super(application);
        retoRepository = new RetoRepository(application);
        pokemonRepository = new PokemonRepository();
    }

    public LiveData<GameChallengeUiState> getChallengeState() {
        return challengeState;
    }

    public void cargarRetoConPokemonAleatorio() {
        challengeState.postValue(GameChallengeUiState.loading());

        executor.execute(() -> {
            Reto reto = retoRepository.obtenerRetoAleatorio();
            if (reto == null) {
                challengeState.postValue(GameChallengeUiState.empty());
                return;
            }

            String pokemonImageUrl = null;
            String errorMessage = null;

            try {
                pokemonImageUrl = pokemonRepository.obtenerImagenAleatoria();
            } catch (Exception exception) {
                errorMessage = "No se pudo cargar la imagen del Pokémon.";
            }

            challengeState.postValue(GameChallengeUiState.content(reto, pokemonImageUrl, errorMessage));
        });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executor.shutdown();
    }
}

File: app/src/main/java/com/example/pico_botella/viewmodel/GameViewModel.kt
package com.example.pico_botella.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.pico_botella.model.GameChallengeUiState
import com.example.pico_botella.repository.PokemonRepository
import com.example.pico_botella.repository.RetoRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject
import kotlin.random.Random

@HiltViewModel
class GameViewModel @Inject constructor(
    private val retoRepository: RetoRepository
) : ViewModel() {

    private val pokemonRepository = PokemonRepository()
    private val _challengeState = MutableLiveData<GameChallengeUiState>()
    val challengeState: LiveData<GameChallengeUiState> = _challengeState

    fun cargarRetoConPokemonAleatorio() {
        _challengeState.value = GameChallengeUiState.loading()

        viewModelScope.launch {
            // Obtener todos los retos de Firestore y elegir uno al azar localmente
            // (Firestore no tiene un "obtener uno al azar" nativo eficiente para listas pequeñas)
            val retos = withContext(Dispatchers.IO) {
                retoRepository.getAllRetos().first()
            }

            if (retos.isEmpty()) {
                _challengeState.value = GameChallengeUiState.empty()
                return@launch
            }

            val retoAleatorio = retos[Random.nextInt(retos.size)]

            var errorMessage: String? = null
            val pokemonImageUrl = withContext(Dispatchers.IO) {
                try {
                    pokemonRepository.obtenerImagenAleatoria()
                } catch (_: Exception) {
                    errorMessage = "No se pudo cargar la imagen del Pokemon."
                    null
                }
            }

            _challengeState.value = GameChallengeUiState.content(retoAleatorio, pokemonImageUrl, errorMessage)
        }
    }
}
